code/modules/power/supermatter/supermatter_process.dm

```diff
@@ -138,6 +139,95 @@
+			var/causality_filter = filter(type="layer", icon = causality_field, flags = FILTER_OVERLAY)
+			set_light(initial(light_range) + clamp(damage*power, 50, 500), 3, (gasmix_power_ratio > 0.8 ? SUPERMATTER_RED : SUPERMATTER_COLOUR), TRUE)
+			filters_to_add |= list(rays_filter, causality_filter)
+
+	switch(check_special_delamination())
+
+		if(CASCADE_DELAMINATION)
+			set_light(initial(light_range) + clamp(damage*power, 50, 500), 3, SUPERMATTER_CASCADE_COLOUR, TRUE)
+			filters_to_add |= filter(type="rays", size = clamp((damage/100)*power, 50, 125), color = SUPERMATTER_CASCADE_COLOUR, factor = clamp(damage/300, 1, 30), density = clamp(damage/5, 12, 200))
+
+
+		if(SINGULARITY_DELAMINATION)
+			var/rays_filter = filter(type="rays", size = clamp((damage/100)*power, 50, 125), color = SUPERMATTER_SINGULARITY_RAYS_COLOUR, factor = clamp(damage/300, 1, 30), density = clamp(damage/5, 12, 200))
+			var/outline_filter = filter(type = "outline", size = 1, color = SUPERMATTER_SINGULARITY_LIGHT_COLOUR)
+
+			if(!warp)
+				warp = new(src)
+				vis_contents += warp
+				animate(warp, time = 20*0.1*3, transform = matrix().Scale(0.5,0.5), loop = 1)
+				animate(time = 20*0.1*7, transform = matrix(), loop = 1)
+
+
+			set_light(initial(light_range) + clamp(damage*power, 10, 50), 3, SUPERMATTER_SINGULARITY_LIGHT_COLOUR, TRUE)
+			filters_to_add |= list(rays_filter, outline_filter) // The Singularity is coming!
+			if(final_countdown)
+				var/icon/causality_field = new/icon('icons/obj/supermatter.dmi', "causality_field", frame = rand(1,4))
+				var/causality_filter = filter(type="layer", icon = causality_field, flags = FILTER_OVERLAY)
+				var/icon/singulo = new/icon('icons/effects/96x96.dmi', "singularity_s3", frame = rand(1,8))
+				var/singulo_filter = filter(type="layer", icon = singulo, flags = FILTER_OVERLAY)
+				filters_to_add |= list(causality_filter, singulo_filter)
+
+		if(TESLA_DELAMINATION)
+			var/rays_filter = filter(type="rays", size = clamp((damage/100)*power, 50, 125), color = SUPERMATTER_TESLA_COLOUR, factor = clamp(damage/300, 1, 30), density = clamp(damage/5, 12, 200))
+			var/icon/ball = new/icon('icons/obj/tesla_engine/energy_ball.dmi', "energy_ball", frame = rand(1,12))
+			var/tesla_filter = filter(type="layer", icon = ball, flags = FILTER_UNDERLAY)
+			set_light(initial(light_range) + clamp(damage*power, 50, 500), 3, SUPERMATTER_TESLA_COLOUR, TRUE)
+			filters_to_add |= list(rays_filter, tesla_filter)
+			if(final_countdown)
+				var/icon/causality_field = new/icon('icons/obj/supermatter.dmi', "causality_field", frame = rand(1,4))
+				var/causality_filter = filter(type="layer", icon = causality_field, flags = FILTER_OVERLAY)
+				filters_to_add |= causality_filter
+	filters = filters_to_add
+
+/// Sucks ALL the objects into the supermatter. Really cool with the singularity.
+/obj/machinery/power/supermatter_crystal/proc/supermatter_pull_delamination(turf, radius = 20, singularity)
+	turf = get_turf(loc)
+	if(!turf)
+		return
```

Comment:
the turf var here absolutely sucks, why is it an arg either? It's not like we ever use it.